File: app/src/main/java/org/vosk/demo/VoskActivity.java
package org.vosk.demo;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.method.ScrollingMovementMethod;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;

import com.google.android.material.button.MaterialButton;

import org.vosk.LibVosk;
import org.vosk.LogLevel;
import org.vosk.Model;
import org.vosk.Recognizer;
import org.vosk.android.RecognitionListener;
import org.vosk.android.SpeechService;
import org.vosk.android.StorageService;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class VoskActivity extends Activity implements RecognitionListener {

    static private final int STATE_START = 0;
    static private final int STATE_READY = 1;
    static private final int STATE_DONE = 2;
    static private final int STATE_MIC = 3;

    /* Used to handle permission request */
    private static final int PERMISSIONS_REQUEST_RECORD_AUDIO = 1;

    private Map<String, Model> loadedModels = new HashMap<>();
    private List<String> availableModelNames = new ArrayList<>();
    private String selectedModelName = "en-us"; // Default model selection

    private SpeechService speechService;
    private TextView resultView;

    @Override
    public void onCreate(Bundle state) {
        super.onCreate(state);
        setContentView(R.layout.main);

        // Setup layout
        resultView = findViewById(R.id.result_text);
        setUiState(STATE_START);

        findViewById(R.id.recognize_mic).setOnClickListener(view -> recognizeMicrophone());
        ((MaterialButton) findViewById(R.id.pause)).setOnClickListener(view -> {
            MaterialButton pauseButton = (MaterialButton) view;
            pause(pauseButton.isChecked());
        });

        LibVosk.setLogLevel(LogLevel.INFO);

        // Check if user has given permission to record audio, init the models after permission is granted
        int permissionCheck = ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.RECORD_AUDIO);
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO}, PERMISSIONS_REQUEST_RECORD_AUDIO);
        } else {
            initModels();
        }
    }

    private void initModels() {
        try {
            String[] modelDirs = getAssets().list(""); // List all assets
            if (modelDirs != null) {
                for (String dir : modelDirs) {
                    if (dir.startsWith("model-")) {
                        String modelName = dir.replace("model-", ""); // Extract language name from directory
                        loadModel(dir, modelName);
                    }
                }
            }
        } catch (IOException e) {
            setErrorState("Error loading models: " + e.getMessage());
        }
    }

    private void loadModel(String assetDir, String modelName) {
        StorageService.unpack(this, assetDir, modelName,
                (model) -> {
                    loadedModels.put(modelName, model);
                    availableModelNames.add(modelName);
                    checkIfModelsReady();
                },
                (exception) -> setErrorState("Failed to unpack the model " + modelName + ": " + exception.getMessage()));
    }

    private void checkIfModelsReady() {
        if (!loadedModels.isEmpty()) {
            setUiState(STATE_READY);
            populateModelComboBox();
        }
    }

    private void populateModelComboBox() {
        Spinner modelSpinner = findViewById(R.id.model_spinner);

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, availableModelNames);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        modelSpinner.setAdapter(adapter);

        modelSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parentView, View selectedItemView, int position, long id) {
                selectedModelName = availableModelNames.get(position); // Save selected model name
            }

            @Override
            public void onNothingSelected(AdapterView<?> parentView) {
                // Default to first model if none is selected
                selectedModelName = availableModelNames.get(0);
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSIONS_REQUEST_RECORD_AUDIO) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                // Recognizer initialization is a time-consuming and it involves IO,
                // so we execute it in async task
                initModels();
            } else {
                finish();
            }
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (speechService != null) {
            speechService.stop();
            speechService.shutdown();
        }
    }

    @Override
    public void onResult(String hypothesis) {
        resultView.append(hypothesis + "\n");
    }

    @Override
    public void onFinalResult(String hypothesis) {
        resultView.append(hypothesis + "\n");
        setUiState(STATE_DONE);
    }

    @Override
    public void onPartialResult(String hypothesis) {
        resultView.append(hypothesis + "\n");
    }

    @Override
    public void onError(Exception e) {
        setErrorState(e.getMessage());
    }

    @Override
    public void onTimeout() {
        setUiState(STATE_DONE);
    }

    private void setUiState(int state) {
        Spinner modelSpinner = findViewById(R.id.model_spinner);

        switch (state) {
            case STATE_START:
                resultView.setText(R.string.preparing);
                resultView.setMovementMethod(new ScrollingMovementMethod());
                findViewById(R.id.recognize_mic).setEnabled(false);
                modelSpinner.setEnabled(false);
                findViewById(R.id.pause).setEnabled((false));
                break;
            case STATE_READY:
                resultView.setText(R.string.ready);
                findViewById(R.id.recognize_mic).setEnabled(true);
                modelSpinner.setEnabled(true);
                findViewById(R.id.pause).setEnabled((false));
                break;
            case STATE_DONE:
                ((Button) findViewById(R.id.recognize_mic)).setText(R.string.listen_microphone);
                findViewById(R.id.recognize_mic).setEnabled(true);
                modelSpinner.setEnabled(true);
                findViewById(R.id.pause).setEnabled((false));
                break;
            case STATE_MIC:
                ((Button) findViewById(R.id.recognize_mic)).setText(R.string.stop_microphone);
                resultView.setText(getString(R.string.say_something));
                findViewById(R.id.recognize_mic).setEnabled(true);
                modelSpinner.setEnabled(false);
                findViewById(R.id.pause).setEnabled((true));
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + state);
        }
    }

    private void setErrorState(String message) {
        resultView.setText(message);
        ((Button) findViewById(R.id.recognize_mic)).setText(R.string.listen_microphone);
        findViewById(R.id.recognize_mic).setEnabled(false);
        findViewById(R.id.model_spinner).setEnabled(false);
    }

    private void recognizeMicrophone() {
        if (speechService != null) {
            setUiState(STATE_DONE);
            speechService.stop();
            speechService = null;
        } else {
            setUiState(STATE_MIC);
            try {
                Recognizer rec = new Recognizer(loadedModels.get(selectedModelName), 16000.0f);
                speechService = new SpeechService(rec, 16000.0f);
                speechService.startListening(this);
            } catch (IOException e) {
                setErrorState(e.getMessage());
            }
        }
    }

    private void pause(boolean checked) {
        if (speechService != null) {
            speechService.setPause(checked);
        }
    }
}

File: app/src/main/java/org/vosk/demo/VoskActivity.kt
package org.vosk.demo

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.speech.tts.TextToSpeech
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.google.accompanist.permissions.rememberPermissionState
import org.json.JSONException
import org.json.JSONObject
import org.vosk.LibVosk
import org.vosk.LogLevel
import org.vosk.Model
import org.vosk.Recognizer
import org.vosk.android.RecognitionListener
import org.vosk.android.SpeechService
import org.vosk.android.StorageService
import org.vosk.demo.ui.theme.VoskTheme
import java.io.IOException
import java.util.*

class VoskActivity : ComponentActivity(), RecognitionListener {
    private var currentModel: Model? = null
    private var speechService: SpeechService? = null
    private var tts: TextToSpeech? = null

    private val availableModelNames = mutableStateListOf<String>()
    private var recognizedText by mutableStateOf("")
    private var partialResult by mutableStateOf("")
    private var uiState by mutableStateOf(UiState.START)
    private var selectedModelName by mutableStateOf("en-us")
    private var errorMessage by mutableStateOf<String?>(null)
    private var isDropdownExpanded by mutableStateOf(false)
    private var isPaused by mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        LibVosk.setLogLevel(LogLevel.INFO)

        // Initialize TTS
        tts = TextToSpeech(this) { status ->
            if (status != TextToSpeech.SUCCESS) {
                errorMessage = "TTS initialization failed"
            }
        }

        setContent {
            VoskTheme {
                MainScreen()
            }
        }

        // Check permissions and initialize
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
            != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                PERMISSIONS_REQUEST_RECORD_AUDIO
            )
        } else {
            initModels()
        }
    }

    @OptIn(ExperimentalPermissionsApi::class, ExperimentalMaterial3Api::class)
    @Composable
    private fun MainScreen() {
        val audioPermissionState = rememberPermissionState(
            Manifest.permission.RECORD_AUDIO
        )

        Surface(
            modifier = Modifier.fillMaxSize(),
            color = MaterialTheme.colorScheme.background
        ) {
            Card(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp),
                elevation = CardDefaults.cardElevation(defaultElevation = 8.dp)
            ) {
                Column(
                    modifier = Modifier
                        .padding(16.dp)
                        .verticalScroll(rememberScrollState()),
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    // Logo
                    Image(
                        painter = painterResource(id = R.drawable.logo),
                        contentDescription = stringResource(R.string.app_logo),
                        modifier = Modifier
                            .size(230.dp, 86.dp)
                            .padding(bottom = 16.dp)
                    )

                    // Language selector
                    Text(
                        text = stringResource(R.string.select_model),
                        style = MaterialTheme.typography.bodyLarge,
                        modifier = Modifier.padding(bottom = 8.dp)
                    )

                    ExposedDropdownMenuBox(
                        expanded = isDropdownExpanded,
                        onExpandedChange = { isDropdownExpanded = it },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        OutlinedTextField(
                            value = selectedModelName,
                            onValueChange = {},
                            readOnly = true,
                            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = isDropdownExpanded) },
                            modifier = Modifier
                                .menuAnchor()
                                .fillMaxWidth(),
                            enabled = uiState != UiState.START && uiState != UiState.MIC
                        )

                        ExposedDropdownMenu(
                            expanded = isDropdownExpanded,
                            onDismissRequest = { isDropdownExpanded = false }
                        ) {
                            availableModelNames.forEach { model ->
                                DropdownMenuItem(
                                    text = { Text(model) },
                                    onClick = {
                                        selectedModelName = model
                                        isDropdownExpanded = false
                                        loadAndSwitchModel(model)
                                    }
                                )
                            }
                        }
                    }

                    // Microphone button
                    OutlinedButton(
                        onClick = { recognizeMicrophone() },
                        enabled = uiState != UiState.START,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp)
                    ) {
                        Icon(
                            painter = painterResource(id = R.drawable.baseline_mic_24),
                            contentDescription = null,
                            modifier = Modifier.padding(end = 8.dp)
                        )
                        Text(
                            text = when (uiState) {
                                UiState.MIC -> stringResource(R.string.stop_microphone)
                                else -> stringResource(R.string.listen_microphone)
                            }
                        )
                    }

                    // Pause button
                    OutlinedButton(
                        onClick = {
                            isPaused = !isPaused
                            pause(isPaused)
                        },
                        enabled = uiState == UiState.MIC,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp)
                    ) {
                        Icon(
                            painter = painterResource(
                                id = if (isPaused) R.drawable.baseline_play_arrow_24
                                else R.drawable.baseline_pause_24
                            ),
                            contentDescription = null,
                            modifier = Modifier.padding(end = 8.dp)
                        )
                        Text(
                            text = if (isPaused) stringResource(R.string.continueRecognition)
                            else stringResource(R.string.pause)
                        )
                    }

                    // Speak button
                    OutlinedButton(
                        onClick = { speakRecognizedText() },
                        enabled = uiState == UiState.DONE && recognizedText.isNotEmpty(),
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp)
                    ) {
                        Icon(
                            painter = painterResource(id = R.drawable.baseline_volume_up_24),
                            contentDescription = null,
                            modifier = Modifier.padding(end = 8.dp)
                        )
                        Text(stringResource(R.string.speak_recognized_text))
                    }

                    // Results text
                    Card(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(top = 16.dp),
                        elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
                    ) {
                        Text(
                            text = errorMessage ?: when (uiState) {
                                UiState.START -> stringResource(R.string.preparing)
                                UiState.READY -> stringResource(R.string.ready)
                                UiState.MIC -> partialResult.ifEmpty { stringResource(R.string.say_something) }
                                UiState.DONE -> recognizedText.ifEmpty { stringResource(R.string.ready) }
                            },
                            modifier = Modifier.padding(16.dp),
                            style = MaterialTheme.typography.bodyLarge
                        )
                    }
                }
            }
        }
    }

    private fun initModels() {
        try {
            val modelDirs = assets.list("") // List all assets
            if (modelDirs != null) {
                for (dir in modelDirs) {
                    if (dir.startsWith("model-")) {
                        val modelName = dir.replace("model-", "")
                        availableModelNames.add(modelName)
                    }
                }
                if (availableModelNames.isNotEmpty()) {
                    selectedModelName = availableModelNames[0]
                    loadAndSwitchModel(selectedModelName)
                }
            }
        } catch (e: IOException) {
            errorMessage = "Error retrieving model directories: ${e.message}"
        }
    }

    private fun loadAndSwitchModel(modelName: String) {
        uiState = UiState.START
        errorMessage = null

        // Unload the previous model if it exists
        currentModel?.close()

        // Load the new model from assets
        val modelDir = "model-$modelName"
        StorageService.unpack(
            this,
            modelDir,
            modelName,
            { model: Model? ->
                currentModel = model
                // Set TTS locale based on the loaded model
                val ttsLocale = getTtsLocaleForModel(modelName)
                if (tts?.isLanguageAvailable(ttsLocale) == TextToSpeech.LANG_AVAILABLE) {
                    tts?.language = ttsLocale
                } else {
                    errorMessage = "TTS language not supported for $modelName"
                }
                uiState = UiState.READY
            },
            { exception: IOException ->
                errorMessage = "Failed to load model $modelName: ${exception.message}"
            }
        )
    }

    private fun recognizeMicrophone() {
        if (speechService != null) {
            uiState = UiState.DONE
            speechService?.stop()
            speechService = null
        } else {
            uiState = UiState.MIC
            try {
                currentModel?.let { model ->
                    val rec = Recognizer(model, 16000.0f)
                    speechService = SpeechService(rec, 16000.0f)
                    speechService?.startListening(this)
                } ?: run {
                    errorMessage = "No model loaded!"
                }
            } catch (e: IOException) {
                errorMessage = e.message
            }
        }
    }

    private fun pause(checked: Boolean) {
        speechService?.setPause(checked)
    }

    private fun speakRecognizedText() {
        if (!recognizedText.isBlank()) {
            tts?.speak(recognizedText, TextToSpeech.QUEUE_FLUSH, null, null)
        }
    }

    override fun onResult(hypothesis: String) {
        try {
            val jsonResult = JSONObject(hypothesis)
            val text = jsonResult.optString("text", "")
            if (text.isNotEmpty()) {
                recognizedText = text
                partialResult = ""
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    override fun onPartialResult(hypothesis: String) {
        try {
            val jsonPartialResult = JSONObject(hypothesis)
            val partial = jsonPartialResult.optString("partial", "")
            if (partial.isNotEmpty()) {
                partialResult = partial
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

    override fun onFinalResult(hypothesis: String) {
        uiState = UiState.DONE
    }

    override fun onError(e: Exception) {
        errorMessage = e.message
    }

    override fun onTimeout() {
        uiState = UiState.DONE
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSIONS_REQUEST_RECORD_AUDIO) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                initModels()
            } else {
                finish()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        tts?.run {
            stop()
            shutdown()
        }

        speechService?.run {
            stop()
            shutdown()
        }

        currentModel?.close()
    }

    private fun getTtsLocaleForModel(modelName: String): Locale = when (modelName) {
        "english" -> Locale.ENGLISH
        "hindi" -> Locale("hi")
        "kazakh" -> Locale("kk")
        "polish" -> Locale("pl")
        "russian" -> Locale("ru")
        "turkish" -> Locale("tr")
        "ukrainian" -> Locale("uk")
        "uzbek" -> Locale("uz")
        else -> Locale.ENGLISH
    }

    enum class UiState {
        START, READY, DONE, MIC
    }

    companion object {
        private const val PERMISSIONS_REQUEST_RECORD_AUDIO = 1
    }
}